Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -24,28 +26,85 @@ namespace BuildXL.Cache.ContentStore.Service.Grpc
     /// </summary>
     public sealed class GrpcCopyClient : IShutdown<BoolResult>
     {
+        // Link to DistributedContentStoreSettings.MaxConcurrentCopyOperations
+        private const int _maxClientCount = 512;
+
+        private static readonly object lockObject = new object();
+        internal static readonly ConcurrentDictionary<(string, int, bool), GrpcCopyClient> _clientDict = new ConcurrentDictionary<(string, int, bool), GrpcCopyClient>(Environment.ProcessorCount, _maxClientCount);

```

Comment:
>_clientDict [](start = 91, length = 11)

Just checking:
Since you use Tuple<,,> you can't control the casing of the string value. Which seems to be the host. Is there any chance there could be a variation in the host (i.e. IP vs DNS, IPv4 vs IPv4 or even casing differences in case of DNS name)?

Reply:
Great point! This should be casing agnostic, and I don't think we should worry about having multiple hosts which evaluate to the same thing. All GrpcCopyClient constructors use the same method to construct the host.

---
In reply to: [289191783](https://github.com/microsoft/BuildXL/pull/391#discussion_r289191783) [](ancestors = 289191783)